core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java

```diff
@@ -0,0 +1,315 @@
+        final AtomicBoolean delegationToMaster = new AtomicBoolean();
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
+            @Override
+            protected void processBeforeDelegationToMaster(Request request, ClusterState state) {
+                logger.debug("Delegation to master called");
+                delegationToMaster.set(true);
+            }
+        }.execute(request, listener);
+
+        assertTrue("processBeforeDelegationToMaster not called", delegationToMaster.get());
+        assertThat(transport.capturedRequests().length, equalTo(1));
+        CapturingTransport.CapturedRequest capturedRequest = transport.capturedRequests()[0];
+        assertTrue(capturedRequest.node.isMasterNode());
+        assertThat(capturedRequest.request, equalTo(request));
+        assertThat(capturedRequest.action, equalTo("testAction"));
+
+        Response response = new Response();
+        transport.handleResponse(capturedRequest.requestId, response);
+        assertThat(listener.get(), equalTo(response));
+    }
+
+    @Test
+    public void testDelegateToFailingMaster() throws ExecutionException, InterruptedException {
+        boolean failsWithConnectTransportException = randomBoolean();
+        Request request = new Request().masterNodeTimeout(TimeValue.timeValueSeconds(failsWithConnectTransportException ? 60 : 0));
+        clusterService.setState(ClusterStateCreationUtils.state(localNode, remoteNode, allNodes));
+
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        final AtomicBoolean delegationToMaster = new AtomicBoolean();
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
+            @Override
+            protected void processBeforeDelegationToMaster(Request request, ClusterState state) {
+                logger.debug("Delegation to master called");
+                delegationToMaster.set(true);
+            }
+        }.execute(request, listener);
+
+        assertTrue("processBeforeDelegationToMaster not called", delegationToMaster.get());
+        assertThat(transport.capturedRequests().length, equalTo(1));
+        CapturingTransport.CapturedRequest capturedRequest = transport.capturedRequests()[0];
+        assertTrue(capturedRequest.node.isMasterNode());
+        assertThat(capturedRequest.request, equalTo(request));
+        assertThat(capturedRequest.action, equalTo("testAction"));
+
+        if (failsWithConnectTransportException) {
+            transport.handleResponse(capturedRequest.requestId, new ConnectTransportException(remoteNode, "Fake error"));
+            assertFalse(listener.isDone());
+            clusterService.setState(ClusterStateCreationUtils.state(localNode, localNode, allNodes));
+            listener.get();
+        } else {
+            Throwable t = new Throwable();
+            transport.handleResponse(capturedRequest.requestId, t);
+            try {
```

Comment:
can we check here if the listener is done?  (just checking if I got it right this time :))